src/modules/keycloak-administration/domain/keycloak-user.service.ts

```diff
@@ -273,4 +273,135 @@ export class KeycloakUserService {
 
         return { ok: true, value: userDo };
     }
+
+    public async assignRealmRolesToUser(usernameId: string, roleNames: string[]): Promise<Result<void, DomainError>> {
+        const kcAdminClientResult: Result<KeycloakAdminClient, DomainError> =
+            await this.kcAdminService.getAuthedKcAdminClient();
+
+        if (!kcAdminClientResult.ok) {
+            return kcAdminClientResult;
+        }
+        // this function returns the findby user
+        const userResult: Result<User<true>, DomainError> = await this.findById(usernameId);
+        if (!userResult.ok) {
+            return userResult;
+        }
+
+        const userId: string = userResult.value.id;
+
+        try {
+            const allRoles: RoleRepresentation[] = await kcAdminClientResult.value.roles.find();
+            // return all roles instead of calling the DB multiple times KC does not support finding all the roles with one call
+            const rolesToAssign: RoleRepresentation[] = allRoles.filter((role: RoleRepresentation) =>
+                roleNames.some((roleName: string) => role.name === roleName),
+            );
+
+            const validRoles: RoleRepresentation[] = rolesToAssign.filter(
+                (role: RoleRepresentation | undefined): role is RoleRepresentation =>
+                    role !== undefined && role.id !== undefined && role.name !== undefined,
+            );
+
+            if (validRoles.length === 0) {
+                return {
+                    ok: false,
+                    error: new EntityNotFoundError(`No valid roles found for the provided role names`),
+                };
+            }
+            const userCurrentRoles: RoleRepresentation[] = await kcAdminClientResult.value.users.listRealmRoleMappings({
+                id: userId,
+            });
+
+            const newRolesToAssign: RoleRepresentation[] = validRoles.filter(
+                (role: RoleRepresentation) =>
+                    !userCurrentRoles.some((userRole: RoleRepresentation) => userRole.id === role.id),
+            );
+
+            if (newRolesToAssign.length === 0) {
+                return { ok: true, value: undefined };
+            }
+
+            const roleMappings: {
+                id: string;
+                name: string;
+            }[] = newRolesToAssign.map((role: RoleRepresentation) => ({
+                id: role.id!,
```

Comment:
This is unnecessary mapping, because id and name already exist in RoleRepresentation.